modules/src/ics02_client/msgs.rs

```diff
@@ -1,25 +1,25 @@
-use super::client_type::ClientType;
-use super::header::Header;
-use super::state::ConsensusState;
+//! These are definitions of messages that a relayer submits to a chain. Specific implementations of
+//! these messages can be found, for instance, in ICS 07 for Tendermint-specific chains. A chain
+//! handles these messages in two layers: first with the general ICS 02 client handler, which
+//! subsequently calls into the chain-specific (e.g., ICS 07) client handler. See:
+//! https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics#create.
+
+use crate::ics02_client::client_def::ClientDef;
+use crate::ics02_client::client_type::ClientType;
 use crate::ics24_host::identifier::ClientId;
-use crate::tx_msg::Msg;
 
-pub trait MsgUpdateClient
-where
-    Self: Msg,
-{
-    type Header: Header;
-    fn client_id(&self) -> &ClientId;
-    fn header(&self) -> &Self::Header;
+/// A type of message that triggers the creation of a new on-chain (IBC) client.
+#[derive(Clone, Debug)]
+pub struct MsgCreateClient<CD: ClientDef> {
+    pub client_id: ClientId,
+    pub client_type: ClientType,
+    pub consensus_state: CD::ConsensusState,
 }
 
-pub trait MsgCreateClient
-where
-    Self: Msg,
-{
-    type ConsensusState: ConsensusState;
-
-    fn client_id(&self) -> &ClientId;
-    fn client_type(&self) -> ClientType;
-    fn consensus_state(&self) -> Self::ConsensusState;
+/// A type of message that triggers the update of an on-chain (IBC) client with new headers.
+#[derive(Clone, Debug)]
+pub struct MsgUpdateClient<CD: ClientDef> {
+    pub client_id: ClientId,
+    pub header: CD::Header,
 }
```

Comment:
The structure of `MsgCreateClient` is different depending on the client type. Same for `MsgUpdateClient`. For example for `tendermint` client, the `MsgCreateClient` includes fields like `trusting_period`, `unbonding_period`, etc. While for `loopback` (not implemented) or other client types, these are not present while others might be. Not sure how we can implement the ics02 messages without traits.

Reply:
Actually, I think this will work if we create the concrete `ClientState`, `ConsensusState`, etc, from the wire/ raw message and then we call `dispatch` -> `process`. So some of the logic will be in this pre-processing step. I will give it a try.